### ISSUE TITLE
Integration with Bookshelf / Knex

### ISSUE THREAD

**Issue by eightyfive** [krakenjs/kraken-js#408]
Hi,

I was wondering what is the best way to integrate Bookshelf with Kraken.

Given the following [file structure](http://davidhunt.io/making-a-new-node-js-app-feel-more-like-rails-part-1):
```
app/
   services/
      bookshelf.js
      index.js
   models/
      user.js
```

My problem is I don't know how to access the `confit` object in `services/bookshelf.js`:
```
// **
// How to access `app.kraken` here ?
// **
var knex = require('knex')(app.kraken.get('database'));

var bookshelf = module.exports = require('bookshelf')(knex);
```

For info, the [Ghost project](https://github.com/TryGhost/Ghost) is using the same strategy: Accessing some config object (analog to `confit`) [from the bookshelf instance file/module](https://github.com/TryGhost/Ghost/blob/master/core/server/models/base/index.js#L10).

Should I change my strategy and set `bookshelf` instance on `app` ?
```
app.use(kraken({
   onconfig: function(config, next) {
        var bookshelf = ... bookshelf instance with newly created config ...
        app.set('bookshelf', bookshelf);

        next(null, config);
   }
})
...
```

But then again, how do I access `bookshelf` instance in my Models:
```
// models/user.js

var app = require('../index); // feels wrong...
var bookshelf = app.get('bookshelf');

var User = module.exports = bookshelf.Model.extend({
...
```

Accessing `app` in the Model layer feels wrong and what about if I'm trying to load my Models without launching the `app`? (like for examples when [running migrations](http://knexjs.org/#Migrations)).

I think my initial strategy is best but how to resolve the `confit.get('database')` dependency??

Sources:
- Recommended way to [bootstrap Bookshelf](http://bookshelfjs.org/#installation)
- [services/bookshelf strategy](http://davidhunt.io/making-a-new-node-js-app-feel-more-like-rails-part-1) and [how to use it in Models](http://davidhunt.io/making-a-new-node-js-app-feel-more-like-rails-part-2)

**Comment by jasisk** [comment]
This is a great question and one we see quite often.

You're absolutely right in the concerns you expressed. The design decisions we made were, in part, to avoid some of the common problems we were seeing when folks start (mis)using global instances, and configuration was a big example of that. The tradeoff, however, is that it can be difficult to access an object that you _actually_ want to be global using the same patterns we've established.

Now, the answer you don't want: there isn't a single pattern I'd _necessarily_ promote over any other.

As for the ghost approach: they're simply leveraging the module caching mechanism as a way to get a singleton. It might _feel_ nicer than `globals.config` but it is, essentially, the same thing.

A few options for you:
1. set a `global` in `onconfig` for your DB connection.
2. write a middleware thunk (to be configured in the `middleware` section of your config) that works like a cache. Something like:
``` js
module.exports = function thunk(dbArgs, modelPath) {
  var models = {};
  // ... for example but this could instead be part of your config ...
  var connection = db.connect(dbArgs);

  return function initModels(req, res, next) {
    req.getModel = function getModel(name) {
      var model;

      if (models[name]) { return models[name]; }

      // ... model not yet initialized so let's do that ...
      model = tryRequire(path.join(modelPath, name));

      if (model) {
        // note that `req.app.kraken` _is_ your config object.
        models[name] = model.init(connection, req.app.kraken);
      }
    };

    next();
  };
};
```
Then you could easily call `req.getModel('someModel')` and you'd be all set. Yes, you should be cautious about hanging properties off `req`, but it's an established pattern and it is a handy ephemeral object tied to the request. Other attempts to create similar stores are often fraught with caveats.

3. initialize your models somewhere (again, `onconfig` is a good place) and pass in the connection stored in the config in each request handler (keeping in mind, again, that the config can be found at `req.app.kraken`).

There really isn't a single right answer but hopefully this gives you a couple of ideas you feel more comfortable with.

**Comment by qlonik** [comment]
Hello, I would like to share one approach that our team found out while working on our app. Maybe you would spot a problem in our approach, so the feedback is greatly appreciated.

We are using db ORM system that requires the same instance of connection to create models.
E.g. if you are using mongoose, what you could do is in one file you require mongoose module, and establish connection, and in another module, you require mongoose module and create model as if connection is already established. However, our ORM does not support such feature, and we had to 'fake' it.

What we did, is we created file that establishes connection and exports that connection, so all the models could require that one file which always returns already established connection.

```js
// driver.js
var ORM = require('beep-boop-orm');

function init(connection) {
  // perform one time initialization of database, like create models
};

module.exports = function(confit) {
  var connection = new ORM(config.get('db'));
  init(connection);

  module.exports = connection;
  return connection;
};
```

```js
// your file with on config function
var driver = require('./driver');

module.exports = function () {
  return {
    onconfig: function(config, next) {
      var connection = driver(config);

      next(null, config);

      // we also had a problem that we were unsure when connection is truly established,
      // so instead of plain next(null, config), it is possible to do something like this:
      // connection.on('established', function() { next(null, config); });
    }
  };
};
```

And after settings these things up, later in your code (e.g. in controllers), you could require that driver file and it will immediately return connection instance.

Few gotchas with this approach:
 - all later requirements should be happening after kraken is finished starting up. What I mean here is that all later requirements should happen after function was called once.
 - This feels a bit risky because it uses caching feature of require function in node.js. Right now caching feature is stable and specifically described in documentation of node.js, but who knows what they will decide to do with it. Maybe with es6 new imports, they will get rid of node.js's requires all together (I hope they wont, those requires are awesome).

**Comment by eightyfive** [comment]
Thanks for your replies.

I think I also cannot get my head around the concept of needing an _instance_ for **defining** models. Coming from PHP OOP world I've never seen this pattern. Object _definition_ is never tied to an instance!

The only resemblance I can see is when you define a static method on your Models, in order to initialise the db connection on your ActiveRecord pattern. But definition and initialisation of Models are always completely uncorrelated..

The fact to `require('../services').bookshelf` in `models/user.js` definition file feels a bit weird to me.

What if I need to access my Models outside of the express/kraken scope?? (for ex: in a command line "utils" node script or a cron job or..):
- I would require `models/user.js`
- Which will `require('../services').bookshelf` instance..
- Which is tight to express/kraken bootstraping...
- Which I don't want...

What I would want is:
- require `models/user.js`
- **I initialize the db connection myself on my models**, not requiring the service `bookshelf` inside `user.js` definition file...
- I use the models...

This way I can require my Models and use them regardless from where or what I decide to do with them. Say in a command line node script, or sharing models with another non-express app, etc..

So I realise my question was more about how to access the kraken config in a standalone and synchronous manner (so a `confit` question in the end).. In order to have this completely standalone `services/bookshelf.js` file (the instance..).

[ _Note_: The strategy in kraken before `confit` allowed to do that I believe? This strategy was then similar to what is doing Ghost. ]

In that regard @qlonik solution was what I was looking for, but it feels a bit hacky (overriding `module.exports` value the second time, is that sustainable?).

Anyway **my real problem** is why my Models are so tied to this bookshelf instance. It doesn't allow me to require them whenever and however I want.

I think this is a bad design decision from bookshelfjs. It should separate definition and initialisation of Models.

If you have any suggestions, please advice, if not you can close the issue as we're heading a bit off original topic! Sorry for that, was trying to get my head around a couple of things here, I realise.

Thx again.

**Comment by eightyfive** [comment]
FYI

A nice article addressing some of the issues we've talked about here:
[How to structure node.js applications with dependency injection](http://blog.redradix.com/an-architecture-for-complex-node-js-apps-with-dependency-injection/)

And by the way, Bookshelfjs models shouldn't be viewed as equivalent PHP Model `class`, but rather as Model `Services` cause they have a strong dependency on the bookshelf **instance**.

Their name is misleading (at least when coming from conventional OOP).

Closing. Thanks for you help.
